Standalone POS

Talej opened the pull request:
# Description

Implements standalone POS functionality. 

- Changes the enable square POS setting to allow enable Square or Standalone mode.
- Under POS settings adds support for managing categories and products.
- Products can be priced in sats or fiat.
- Adds a standalone POS order creation screen and allows invoicing through the existing payment flow.

** Note: ** I haven't yet tested this on an Android device. Currently without a working device

This pull request is categorized as a:

- [X] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Configuration change
- [ ] Locales update
- [ ] Quality assurance 
- [ ] Other

## Checklist
- [ ] I’ve run `yarn run tsc` and made sure my code compiles correctly
- [X] I’ve run `yarn run lint` and made sure my code didn’t contain any problematic patterns
- [X] I’ve run `yarn run prettier` and made sure my code is formatted correctly
- [X] I’ve run `yarn run test` and made sure all of the tests pass

## Testing

If you modified or added a utility file, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [X] N/A

I have tested this PR on the following platforms (please specify OS version and phone model/VM):

- [ ] Android
- [X] iOS

I have tested this PR with the following types of nodes (please specify node version and API version where appropriate):

- [X] Embedded LND
- [ ] LND (REST)
- [ ] LND (Lightning Node Connect)
- [ ] Core Lightning (c-lightning-REST)
- [ ] LndHub
- [ ] [DEPRECATED] Core Lightning (Spark)
- [ ] [DEPRECATED] Eclair

### Locales
- [X] I’ve added new locale text that requires translations
- [ ] I’m aware that new translations should be made on the Zeus [Transfix page](https://app.transifex.com/ZeusLN/zeus/) and not directly to this repo

### Third Party Dependencies and Packages

- [ ] Contributors will need to run `yarn` after this PR is merged in
- [ ] 3rd party dependencies have been modified:
    * verify that `package.json` and `yarn.lock` have been properly updated
    * verify that dependencies are installed for both iOS and Android platforms

### Other:

- [ ] Changes were made that require an update to the README
- [ ] Changes were made that require an update to onboarding

shubhamkmr04 reviewed (commented):
![Simulator Screenshot - iPhone 14 Plus - 2023-10-12 at 18 45 54](https://github.com/ZeusLN/zeus/assets/50761978/5b5638c4-5ca5-4746-b3e1-3654c864fa9c)
Getting this warning as soon as we open the POS settings

shubhamkmr04 reviewed (commented):
https://github.com/ZeusLN/zeus/assets/50761978/8ed84874-13d6-4286-ac6f-a62bb6664da3

Not able to change the unit to 'sats'

shubhamkmr04 reviewed (commented):
If the Category gets deleted, the product belonging to that Category should also be deleted.

shubhamkmr04 reviewed (commented):
https://github.com/ZeusLN/zeus/assets/50761978/6864f619-9585-46e6-894f-b8d23550f00b

So we add our product to checkout and pay, then delete it from the settings, come to the homepage, and still can see our checked-out price belonging to our deleted product.

kaloudis commented:
> Screen.Recording.2023-10-12.at.10.08.11.PM.mov
> 
> We should have a counter here on this view to know how many products we have selected so far and maybe a way to remove it, too. @kaloudis, what do you think about this?

+1 on having a counter

kaloudis commented:
> Screen.Recording.2023-10-12.at.10.13.21.PM.mov
> 
> So we add our product to checkout and pay, then delete it from the settings, come to the homepage, and still can see our checked-out price belonging to our deleted product.

Deleting a product should probably trigger a cart clear

Talej commented:
> Screen.Recording.2023-10-12.at.10.08.11.PM.mov 
> We should have a counter here on this view to know how many products we have selected so far and maybe a way to remove it, too. @kaloudis, what do you think about this?

I have added an item counter in the "Charge" button.

Other thing I could think to do is add - (n) + in each of the product components. Might not be super user friendly though IMO.

Talej commented:
> ![Simulator Screenshot - iPhone 14 Plus - 2023-10-12 at 18 45 54](https://user-images.githubusercontent.com/50761978/274614164-5b5638c4-5ca5-4746-b3e1-3654c864fa9c.png) Getting this warning as soon as we open the POS settings

This one would've already been in there previously but I have sorted it out by dropping the FlatList

Talej commented:
All of these items and a couple more things I found should be sorted out now.

shubhamkmr04 commented:
> > Screen.Recording.2023-10-12.at.10.08.11.PM.mov
> > We should have a counter here on this view to know how many products we have selected so far and maybe a way to remove it, too. @kaloudis, what do you think about this?
> 
> I have added an item counter in the "Charge" button.
> 
> Other thing I could think to do is add - (n) + in each of the product components. Might not be super user friendly though IMO.

Yup, I understood. @kaloudis, should we do that, or the overall counter would be fine

shubhamkmr04 reviewed (commented):
Should we also remove open orders of deleted products?🤔

shubhamkmr04 reviewed (commented):
https://github.com/ZeusLN/zeus/assets/50761978/bba7ec0d-0144-4d77-9da0-5a80b53577f1


We cannot save a product to the first category in the list; it goes into the uncategorized products. It only happens in the android.

shubhamkmr04 reviewed (commented):
On the **ProductDetails** view, In the delete button, we should have our props                                 `containerStyle` and `titleStyle` set to `themeColor('delete')` to maintain consistency with other views.

shubhamkmr04 commented:
> > Screen.Recording.2023-10-12.at.10.13.21.PM.mov
> > So we add our product to checkout and pay, then delete it from the settings, come to the homepage, and still can see our checked-out price belonging to our deleted product.
> 
> Deleting a product should probably trigger a cart clear

Would it be better to subtract that deleted product's price instead of clearing the whole cart?

Talej commented:
> > > Screen.Recording.2023-10-12.at.10.13.21.PM.mov
> > > So we add our product to checkout and pay, then delete it from the settings, come to the homepage, and still can see our checked-out price belonging to our deleted product.
> > 
> > 
> > Deleting a product should probably trigger a cart clear
> 
> Would it be better to subtract that deleted product's price instead of clearing the whole cart?

I think it would be a pretty rare condition that the user is editing/deleting products at the same time as actually making sales so I feel like just clearing the whole cart should be sufficient. What do you think?

Talej commented:
> Should we also remove open orders of deleted products?🤔

I feel like maybe this would lead to confusion and it should probably be left to the user to either hide/cancel the order or create a new one.

Talej commented:
- Fixed category delete issue with products
- Fixed issue with selecting category for product on android
- Updated styling for delete buttons

Ready for another round!

kaloudis commented:
> > > > Screen.Recording.2023-10-12.at.10.13.21.PM.mov
> > > > So we add our product to checkout and pay, then delete it from the settings, come to the homepage, and still can see our checked-out price belonging to our deleted product.
> > > 
> > > 
> > > Deleting a product should probably trigger a cart clear
> > 
> > 
> > Would it be better to subtract that deleted product's price instead of clearing the whole cart?
> 
> I think it would be a pretty rare condition that the user is editing/deleting products at the same time as actually making sales so I feel like just clearing the whole cart should be sufficient. What do you think?

This should suffice

kaloudis commented:
> > Should we also remove open orders of deleted products?🤔
> 
> I feel like maybe this would lead to confusion and it should probably be left to the user to either hide/cancel the order or create a new one.

can go either way here. I don't think this is a deal breaker either way. maybe something that gets iterated on

shubhamkmr04 commented on views/POS/ProductDetails.tsx:
```diff
@@ -0,0 +1,504 @@
+                    }
+                ];
+                categoryOptions = categoryOptions.concat(...mappedCategories);
+                console.log('options', categoryOptions);
```
Do we need this console, or was it just for debugging?

Talej replied:
Was just for debugging. Have removed now

Talej commented:
> Screen.Recording.2023-10-23.at.4.25.41.PM.mov 
> The screen returns to the top when I select any product, which might affect the user's experience as they would have to scroll down again and again if we have so many products listed.

Having the Charge button up in the Tab Navigator was triggering re-render on every touch. I have restructured it so that doesn't happen any more.

Unfortunately means that I can't use the Tab Navigator space now so we've lost a bit of screen real estate

shubhamkmr04 commented:
![Simulator Screenshot - iPhone 14 Plus - 2023-10-27 at 19 09 20](https://github.com/ZeusLN/zeus/assets/50761978/51a0c0ef-efad-4d9c-b9fb-88c89ca2e707)
@kaloudis, does the button look good in here?

shubhamkmr04 reviewed (commented):
https://github.com/ZeusLN/zeus/assets/50761978/54fddd31-ffa6-4045-a0e7-40b83b608825

The amount is stuck on the price of first product I selected.

Talej commented:
> Screen.Recording.2023-10-27.at.7.24.49.PM.mov 
> The amount is stuck on the price of first product I selected.

Not sure how I missed that one. Nice catch. Fixed now!

shubhamkmr04 reviewed (commented):
I think we shouldn't clear the cart when returning from another screen.

shubhamkmr04 reviewed (commented):
![Simulator Screenshot - iPhone 14 Plus - 2023-10-30 at 14 43 57](https://github.com/ZeusLN/zeus/assets/50761978/9caf3a3f-8ee5-41a5-8b18-2e8db85ffa00)
![Simulator Screenshot - iPhone 14 Plus - 2023-10-30 at 14 43 45](https://github.com/ZeusLN/zeus/assets/50761978/6555343f-ad3b-40b4-b75a-7accf60281fb)
I am getting a weird UI behavior in IOS where the category heading gets stuck on the top until it is scrolled completely.

shubhamkmr04 reviewed (commented):
The latest orders should be listed on the top, just like we get the latest transactions first in the activity view.

shubhamkmr04 reviewed (commented):
https://github.com/ZeusLN/zeus/assets/50761978/3a831eef-2793-4268-84ab-eee0fc76f011

Not able to edit category.

shubhamkmr04 reviewed (commented):
What is the context of the checkbox `Active` while creating the product?

Talej commented:
> Maybe we shouldn't clear the cart when returning from another screen, only when the product is deleted.

It definitely shouldn't be clearing the cart just by changing screens. I have tried to reproduce this a few different ways but it doesn't seem to happen here.

Can you provide any more details?

Talej commented:
> ![Simulator Screenshot - iPhone 14 Plus - 2023-10-30 at 14 43 57](https://user-images.githubusercontent.com/50761978/278999880-9caf3a3f-8ee5-41a5-8b18-2e8db85ffa00.png) ![Simulator Screenshot - iPhone 14 Plus - 2023-10-30 at 14 43 45](https://user-images.githubusercontent.com/50761978/279000124-6555343f-ad3b-40b4-b75a-7accf60281fb.png) I am getting a weird UI behavior in IOS where the category heading gets stuck on the top until it is scrolled completely.

This is just the default SectionList behaviour for iOS. I should be able to disable it

Talej commented:
> The latest orders should be listed on the top, just like we get the latest transactions first in the activity view.

I will update to sort these

Talej commented:
> What is the context of the checkbox `Active` while creating the product?

The main product POS screen should only be showing `Active` products. Will update it

shubhamkmr04 commented:
> > Maybe we shouldn't clear the cart when returning from another screen, only when the product is deleted.
> 
> It definitely shouldn't be clearing the cart just by changing screens. I have tried to reproduce this a few different ways but it doesn't seem to happen here.
> 
> Can you provide any more details?


https://github.com/ZeusLN/zeus/assets/50761978/0232c8d3-8c75-4777-a277-bd9858627ee8

shubhamkmr04 reviewed (commented):
Also, please check the UI for other themes as well. Maybe we need to make `background: none` in some places to make it consistent.

Talej commented:
> Also, please check the UI for other themes as well. Maybe we need to make `background: none` in some places to make it consistent.

For this one I have removed the backgroundColor on the section headers. Haven't made any other changes but I'm open to suggestions.

Talej commented:
Latest changes have been pushed that should cover everything to date. Let me know if there is anything else I've missed.

shubhamkmr04 reviewed (commented):
Let's clear the cart in case we make a selected product inactive.

shubhamkmr04 reviewed (commented):
Also, could you remove the search in all the views when we have nothing to show?

shubhamkmr04 commented on views/Settings/PointOfSale.tsx:
```diff
@@ -415,41 +414,35 @@ export default class PointOfSale extends React.Component<
+                                    containerStyle={{
+                                        borderBottomWidth: 0,
+                                        backgroundColor:
+                                            themeColor('background')
```
Use `backgroundColor: 'none` here.

Talej commented:
In theory everything should be rebased properly and latest feedback all implemented!

kaloudis commented:
Looking great!

kaloudis commented:
Alignment is a little off here on my Android simulator. Could we potentially use the `AmountInput` component here?

<img width="394" alt="Screenshot 2023-12-11 at 15 13 13" src="https://github.com/ZeusLN/zeus/assets/1878621/6dca40d3-3979-4fd4-9478-1d7c7a595e22">

kaloudis commented:
Might be better fit for another PR but I think we should consider adding a keypad pane to the POS for merchants that want to ring up a quick payment without specifying items. Perhaps should be a POS setting.

We should also remove the easter egg to reveal the POS in Settings. It should now be available for everyone to use.

Talej commented:
> Might be better fit for another PR but I think we should consider adding a keypad pane to the POS for merchants that want to ring up a quick payment without specifying items. Perhaps should be a POS setting.
> 
> We should also remove the easter egg to reveal the POS in Settings. It should now be available for everyone to use.

Have removed the easter egg check on the menu directly but left all the easter egg code in place just so it can be used for other options going forward.

Let's do the keypad as a separate PR

kaloudis commented:
<img width="444" alt="Screenshot 2023-12-18 at 19 42 36" src="https://github.com/ZeusLN/zeus/assets/1878621/617ef565-eeb5-4bc9-83d4-1d28e8c34b0b">

hitting an issue here where the `AmountInput` component is pre-filled with value `undefined` and I cannot change the amount

Talej commented:
> <img alt="Screenshot 2023-12-18 at 19 42 36" width="444" src="https://private-user-images.githubusercontent.com/1878621/291442544-617ef565-eeb5-4bc9-83d4-1d28e8c34b0b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5NDY5MTUsIm5iZiI6MTcwMjk0NjYxNSwicGF0aCI6Ii8xODc4NjIxLzI5MTQ0MjU0NC02MTdlZjU2NS1lZWI1LTRiYzktODNkNC0xZDI4ZThjMzRiMGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTlUMDA0MzM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODk4ZWFiYjE1YTE3YWRmZDE2NzBkNTI3ZGQ0ODc0NWM3YmQyY2FlMzRjMjExM2FjMjViZjAzZGJhZDUwY2FlMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.gpYsCnwwSEVxrCKBzRKz_ZNSuqWd_zb0naDqOaDJX2Q">
> hitting an issue here where the `AmountInput` component is pre-filled with value `undefined` and I cannot change the amount

creating a new product or editing an existing?

kaloudis commented:
> > <img alt="Screenshot 2023-12-18 at 19 42 36" width="444" src="https://private-user-images.githubusercontent.com/1878621/291442544-617ef565-eeb5-4bc9-83d4-1d28e8c34b0b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5NDY5MTUsIm5iZiI6MTcwMjk0NjYxNSwicGF0aCI6Ii8xODc4NjIxLzI5MTQ0MjU0NC02MTdlZjU2NS1lZWI1LTRiYzktODNkNC0xZDI4ZThjMzRiMGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTlUMDA0MzM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODk4ZWFiYjE1YTE3YWRmZDE2NzBkNTI3ZGQ0ODc0NWM3YmQyY2FlMzRjMjExM2FjMjViZjAzZGJhZDUwY2FlMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.gpYsCnwwSEVxrCKBzRKz_ZNSuqWd_zb0naDqOaDJX2Q">
> > hitting an issue here where the `AmountInput` component is pre-filled with value `undefined` and I cannot change the amount
> 
> creating a new product or editing an existing?

create a new product

kaloudis commented:
Looking good now! I've also double checked the structure to ensure that fiat prices are recorded properly.

shubhamkmr04 reviewed (commented):
The amount is prefilled with 0 that we can't remove 

<img width="373" alt="Screenshot 2023-12-20 at 12 11 14 PM" src="https://github.com/ZeusLN/zeus/assets/50761978/db12bf4e-fdcf-45ff-b942-3f7d7f3d385c">


<img width="369" alt="Screenshot 2023-12-20 at 12 11 48 PM" src="https://github.com/ZeusLN/zeus/assets/50761978/d7bddf1c-117f-4fdc-ab58-9151d13bf7c0">

Talej commented:
> The amount is prefilled with 0 that we can't remove
> 
> <img alt="Screenshot 2023-12-20 at 12 11 14 PM" width="373" src="https://private-user-images.githubusercontent.com/50761978/291802604-db12bf4e-fdcf-45ff-b942-3f7d7f3d385c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwNTU0NDksIm5iZiI6MTcwMzA1NTE0OSwicGF0aCI6Ii81MDc2MTk3OC8yOTE4MDI2MDQtZGIxMmJmNGUtZmRjZi00NWZmLWI5NDItM2Y3ZDdmM2QzODVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIwVDA2NTIyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyZGRkZDRjMDg3M2NmZWM1ZjFmYWFmNzhkNjcyZDQzZDkyOTc5Njk1MDhiOGE1NDI5M2FkZDUzMjA1NmYzOWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QjpqWhL-0DXhHKhwbWv20XD6kYSZ1H1t1CgXSXqT-SA"> <img alt="Screenshot 2023-12-20 at 12 11 48 PM" width="369" src="https://private-user-images.githubusercontent.com/50761978/291802748-d7bddf1c-117f-4fdc-ab58-9151d13bf7c0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwNTU0NDksIm5iZiI6MTcwMzA1NTE0OSwicGF0aCI6Ii81MDc2MTk3OC8yOTE4MDI3NDgtZDdiZGRmMWMtMTE3Zi00ZmRjLWFiNTgtOTE1MWQxM2JmN2MwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIwVDA2NTIyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4YTQwNjAwNWY5MjA3N2NkMDUyYjhiOTA2N2I0N2M3ZjlhYjMyM2QyZGY3Zjg4YTU3MWU0YmI4NDcxYjU2OTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.CRWz9Hhdy6vbDlNClYVkR0y5SBtK5yE7rBNSQMSSJGA">

Fixed

shubhamkmr04 commented:
Looks fine now.

kaloudis commented:
Code looks good. A few things missing though:

- Ability (or option) to go to order overview before generating invoice (allows user to set tip on this screen and see the total w/ tax) [you can go to this view if you don't pay and click on an order under the Open tab)
- Ability to set tax %
- Ability to disable tips

LMK what you think @Talej. Open to shipping it in beta1 as is and breaking up follow up work in other PRs.

kaloudis commented:
nit: let's also put the Standalone option above Square in the selection menu

kaloudis commented:
I've made the suggested changes myself to the PR:

- Standalone placed above Square in the mode selection dropdown
- Charging an order now takes you to an itemized breakdown instead of directly to the invoice (let's give some thought into if we want to make this configurable)
- Disable tip and conf preferences are now displayed when on the POS settings when in Standalone mode
- New tax percentage setting added to POS settings when in Standalone mode

I also noticed a bug with how itemized products were displayed on the `Order view`: amount displayed was 100x lower. Change can be found here: https://github.com/ZeusLN/zeus/pull/1722/commits/ad4927256c5c34d861fd93567c80be67c2e4d99d#diff-7ffe38a0457034ead72df14de5a05813a6c8a8945e64551040568d72e81a7d5eR330 Just need a sanity check here, but it appears to be rendering properly now.

Talej commented:
> I've made the suggested changes myself to the PR:
> 
> * Standalone placed above Square in the mode selection dropdown
> * Charging an order now takes you to an itemized breakdown instead of directly to the invoice (let's give some thought into if we want to make this configurable)
> * Disable tip and conf preferences are now displayed when on the POS settings when in Standalone mode
> * New tax percentage setting added to POS settings when in Standalone mode
> 
> I also noticed a bug with how itemized products were displayed on the `Order view`: amount displayed was 100x lower. Change can be found here: [ad49272#diff-7ffe38a0457034ead72df14de5a05813a6c8a8945e64551040568d72e81a7d5eR330](https://github.com/ZeusLN/zeus/commit/ad4927256c5c34d861fd93567c80be67c2e4d99d#diff-7ffe38a0457034ead72df14de5a05813a6c8a8945e64551040568d72e81a7d5eR330) Just need a sanity check here, but it appears to be rendering properly now.

This all checks out and looks okay to me!

shubhamkmr04 reviewed (commented):
tACK
